core.network/src/org/netbeans/core/network/utils/SimpleObjCache.java

```diff
@@ -0,0 +1,136 @@
+ * elements will have to be inspected to find the eviction candidate. For this
+ * reason, the class is best suited for smaller caches (say less than 1000
+ * elements).
+ *
+ *
+ * @author lbruun
+ * 
+ * @param <K> key 
+ * @param <V> value
+ */
+public class SimpleObjCache<K,V> {
+
+    private final int maxSize;
+    private final ConcurrentHashMap<K, ValueWrapper> map;
+    
+    /**
+     * Constructs a new cache with {@code maxSize} capacity. When the 
+     * cache is full, the value used/created the furthest in the past will
+     * be evicted to make room for a new element.
+     * 
+     * @param maxSize capacity - any value larger than zero
+     */
+    public SimpleObjCache(int maxSize) {
+        
+        if (maxSize <= 0) {
+            throw new IllegalArgumentException("Cache max size cannot be zero or less");
+        }
+        this.maxSize = maxSize;
+        this.map = new ConcurrentHashMap<>();
+    }
+    
+    /**
+     * Puts a value into the cache mapped to the specified key.
+     * 
+     * @param key key with which the specified value is to be associated
+     * @param value value to be associated with the specified key
+     * @return the previous value associated with key, or null if there 
+     *         was no mapping for the key
+     */
+    public V put(K key, V value) {
+        if (map.size() >= maxSize) {
+            map.remove(findEvictionCandidate());
+        }
+        ValueWrapper existingValue = map.put(key, new ValueWrapper(value));
+        if (existingValue != null) {
+            return existingValue.value;
+        } else {
+            return null;
+        }
+    }
+    
+    /**
+     * Returns the value to which the specified key is mapped, or {@code null}
+     * if this cache contains no mapping for the key.
+     * 
+     * @param key the key whose associated value is to be returned
+     * @return  the value to which the specified key is mapped, or {@code null}
+     *          if this map contains no mapping for the key
+     */
+    public V get(K key) {
+        ValueWrapper wrapper = map.get(key);
+        if (wrapper != null) {
+            wrapper.lastUsed.set(System.currentTimeMillis());
+            return wrapper.value;
+        }
+        return null;
+    }
+    
+    /**
+     * Gets the current number of elements in the cache.
+     * @return 
+     */
+    public int getCacheSize() {
+        return map.size();
+    }
+    
+    /**
+     * Removes all elements from the cache. There is little reason to use
+     * this method as the cache will do its own house keeping.
+     */
+    public void clear() {
+        map.clear();
+    }
+    
+    private K findEvictionCandidate() {
+        // Finds minimum of all timestamps of all elements in the cache.
+        return map.entrySet().stream()
+                .min(Comparator.comparingLong( e -> e.getValue().lastUsed.get()))
+                .get().getKey();
```

Comment:
Why wouldn't it be thread-safe?   map is `ConcurrentHashMap`. I'm confident it is "safe" (as in not throwing errors), but I guess your point is that the stream can (potentially) return a key which has been deleted in the meantime by another thread (two threads both calling `put` at the same time). But this will just mean that one of the threads will call `map.remove()` with a key that no longer exists and that's ok. (it is not an error). I think it is ok that the eviction may theoretically fail to delete something. It will fix itself on the next put operation.

Reply:
concurrentMap.entrySet() is not a copy, it is still backed by the map. So any parallel remove from the origin (concurrent)Map will be reflected in the Set, therefore invalidating Iterators and Streams derived from the entry set. The stream will eventually give a ConcurrentModificationException.

Reply:
@svatoun. Yes, I understand it is still backed by the map. However, the Javadoc for [entrySet](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#entrySet--) says:

> The view's spliterator reports Spliterator.CONCURRENT, Spliterator.DISTINCT, and Spliterator.NONNULL.

(the spliterator is used to create the stream, no?)

Note that this is specific to `ConcurrentHashMap`.  A standard `HashMap` doesn't have these attributes. 

Anyway, the way I understand [Spliterator.CONCURRENT](https://docs.oracle.com/javase/8/docs/api/java/util/Spliterator.html#CONCURRENT) is that is means we'll never get `ConcurrentModificationException`.

I'm far from certain here, so I´m listening!

Reply:
We can add `synchronized` to the `put()` method, even I think it is not needed.  This will for sure be acceptable from a performance point of view. The cache is expected to be used in situations where the there are many, many more reads than writes.

Reply:
Your argument about `Spliterator.CONCURRENT` is in line with the documentation of the ConcurrentHashMap:

> Similarly, Iterators, Spliterators and Enumerations return elements reflecting the state of the hash  table at some point at or since the creation of the iterator/enumeration. They do not throw  ConcurrentModificationException.

so I would remove the synchronized from `evictIfFull`

Reply:
I stand corrected: I didn't consider the origin of the Spliterator. Sorry for the confusion.

Reply:
There's actually still a point to having the `synchronized` there. It not because of thread-safety because the code - even without the `synchronized` is now free from concurrency exceptions, However, it protects against the situation when multiple threads simultaneously arrive at the conclusion that eviction is required, so they all perform an evict, thereby essentially evicting _more_ elements than is strictly needed.